doc/taler-exchange-wire.1
.Dd March 18, 2018
.Dt TALER-EXCHANGE-WIRE 1
.Os
.Sh NAME
.Nm taler-exchange-wire
.Nd create the master-key signed responses to /wire
.Sh SYNOPSIS
.Nm
.Op Fl h | -help
.Op Fl m Ar MASTERKEYFILE | Fl -master= Ns Ar MASTERKEYFILE
.Op Fl v | -version
.Sh DESCRIPTION
.Nm
is used to create the exchange's reply to a /wire request.
It converts the bank details into the appropriate signed response.
This needs to be done using the long-term offline master key.
.Pp
Its options are as follows:
.Bl -tag -width indent
.It Fl h | -help
Print short help on options.
.It Fl m Ar MASTERKEYFILE | Fl -master= Ns Ar MASTERKEYFILE
Specifies the name of the file containing the exchange's master key.
.It Fl v | -version
Print version information.
.El
.Sh SEE ALSO
.Xr taler-exchange-httpd 1 ,
.Xr taler.conf 5 .
.Sh BUGS
Report bugs by using
.Lk https://gnunet.org/bugs/
or by sending electronic mail to
.Aq Mt taler@gnu.org .


ChangeLog
Fri 03 May 2019 05:36:10 PM CEST
    Add support for compressing bodies during HTTP uploads to
    exchange and auditor. -CG

Thu 02 May 2019 09:15:37 PM CEST
    Pass hash of denomination public key in operations from
    clients to exchange instead of the (larger) full public key.
    Breaks protocol compatibility, bumping protocol to v3. -CG

Mon Mar 11 03:24:07 CET 2019
	Completed implementation of #5536 (delete private keys once we
	no longer need them). -CG

Sat Mar  2 19:09:43 CET 2019
	Changing denomination key revocation file format and moving them
	to their own directory (preparations for #5536 resolution). -CG

Sat Feb 23 14:47:12 CET 2019
	Actually disable 100 CONTINUE. -CG

Sun Aug 19 15:15:48 CEST 2018
	Increase various limits and rework transaction scopes to
	improve scalability. -CG

Tue Apr  3 23:29:06 CEST 2018
	Releasing Taler v0.5.0

Thu Jan  4 11:55:41 CET 2018
	Fix issue #5234 (aggregator ignoring refunds).
	Misc. minor fixes to the auditor. -CG

Mon Jan  1 23:15:37 CET 2018
	Add TALER_EXCHANGE_refund2() API call to libtalerexchange. -CG

Thu Dec 14 15:32:50 CET 2017
	Replaced taler-exchange-reservemod tool with new taler-bank-transfer
	tool (#5195).  Removed /admin/add/incoming API, replaced by new
	taler-exchange-wirewatch tooling. (#5077).

Sat Dec  9 15:21:50 CET 2071
	Implement refresh protocol optimization, bumping protocol version
	to v2 (#5178). -CG

Thu Nov  2 17:39:40 CET 2017
	Limit amount values to 2^53 as we always wanted (#5167). -CG

Thu Nov  2 16:50:46 CET 2017
	Fixing serious bug in reserve balance calculations, as
	fees were not always properly included in the summation. -CG

Tue Oct 31 13:43:55 CET 2017
	Extended crypto API to provide planchet generation functions
	explicitly (to be used on withdraw, refresh and tipping).
	Renamed a few API calls to better reflect their functions.
	Changed a few function signatures to use new structs. -CG

Wed Oct 18 15:20:29 CEST 2017
	Releasing Taler v0.4.0. -CG

Tue Oct 17 14:32:46 CEST 2017
	Fix building of /keys response to include full auditor
	signature set instead of partial map. -CG

Mon Oct 16 12:10:35 CEST 2017
	Realize JSON-based report formats from auditor (#4963). -CG

Sun Sep 17 16:46:13 CEST 2017
	Implement /keys cherry picking (#4840). -CG

Wed Sep 13 14:06:44 CEST 2017
	Add argument to TALER_EXCHANGE_check_keys_current()
	to force download even if /keys are still considered
	current. -CG

Fri Jul 14 17:38:54 CEST 2017
	Fix #4751 by not allowing denomination deletion (GC)
	to cascade into reserve_out table (and tolerating such
	deletion failures). This way, denominations will be kept
	around until the last reserve that references them (or
	known_coins derived from them) is closed. Naturally, in
	most case we expect reserves to be closed much faster
	than DKs, so in practice this should rarely apply. -CG

Sun Jun 11 17:03:56 CEST 2017
	Finish implementation and testing of automated refunding
	of expired reserves (#4956). -CG

Tue Jun  6 13:53:34 CEST 2017
	Releasing taler-exchange 0.3.0. -CG

Mon Apr 17 01:29:07 CEST 2017
	Add support for HTTP body compression (#4982). -CG

Mon Mar 20 04:37:46 CET 2017
	Implemented first working version of taler-auditor. -CG

Sat Mar 18 03:44:38 CET 2017
	Add 'wire_out' table to exchange DB to track outgoing
	wire transfers. -CG

Fri Nov 18 18:53:30 CET 2016
	Improved error reporting for bogus wire specifications.
	Releasing taler-exchange 0.2.0. -CG

Thu Nov 17 16:37:22 CET 2016
	Added missing idempotency checks for /refresh/reveal. -CG

Thu Nov 17 11:37:56 CET 2016
	Fixed a few cases of missing database rollbacks, causing the
	exchange to be stuck. -CG

Mon Jun 20 17:40:49 CEST 2016
 	Releasing taler-exchange 0.0.2. -CG

Mon Jun 13 16:35:09 CEST 2016
	Splitting HTTP service to run on different listen
	sockets for /admin and /test-APIs and the normal public
	APIs (#4569). -CG

Thu Jun  9 19:17:02 CEST 2016
	Rename /deposit/wtid to /track/transaction and
	/wire/deposits to /track/transfer. -CG

Wed Jun  8 11:32:41 CEST 2016
	Export libtalerfakebank API so we can use it in testcases for the
	merchant as well. -CG

Tue Jun  7 15:13:46 CEST 2016
        Adding public key of the exchange that was used to sign replies
        to applicable callbacks of libtalerexchange. (This will eventually
        be needed by the merchant's backend.) -CG

Mon Jun  6 23:59:05 CEST 2016
	Minor code cleanup, fixing FTBFS when run with just 'make'
	or when header installed to a location not included.
	Releasing taler-exchange 0.0.1. -CG

Wed Jun  1 17:27:36 CEST 2016
	Releasing taler-exchange 0.0.0. -CG


src/exchange-tools/taler-exchange-wire.c
/*
  This file is part of TALER
  Copyright (C) 2015-2018 Taler Systems SA

  TALER is free software; you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation; either version 3, or (at your option) any later version.

  TALER is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along with
  TALER; see the file COPYING.  If not, see <http://www.gnu.org/licenses/>
*/
/**
 * @file taler-exchange-wire.c
 * @brief Create signed response for /wire requests.
 * @author Christian Grothoff
 */
#include <platform.h>
#include <jansson.h>
#include <gnunet/gnunet_json_lib.h>
#include "taler_crypto_lib.h"
#include "taler_util.h"
#include "taler_wire_lib.h"
#include "taler_json_lib.h"
#include "taler_exchangedb_lib.h"
#include "taler_signatures.h"


/**
 * Filename of the master private key.
 */
static char *masterkeyfile;

/**
 * Private key for signing.
 */
static struct TALER_MasterPrivateKeyP master_priv;

/**
 * Return value from main().
 */
static int global_ret;


/**
 * Function called with information about a wire account.  Signs
 * the account's wire details and writes out the JSON file to disk.
 *
 * @param cls closure
 * @param ai account information
 */
static void
sign_account_data (void *cls,
                   const struct TALER_EXCHANGEDB_AccountInfo *ai)
{
  json_t *wire;
  char *json_out;
  FILE *out;
  int ret;

  if (GNUNET_NO == ai->credit_enabled)
    return;
  if (NULL == ai->wire_response_filename)
  {
    GNUNET_log_config_missing (GNUNET_ERROR_TYPE_ERROR,
                               ai->section_name,
                               "WIRE_RESPONSE");
    global_ret = 1;
    return;
  }
  wire = TALER_JSON_exchange_wire_signature_make (ai->payto_url,
                                                  &master_priv);
  GNUNET_assert (NULL != wire);
  json_out = json_dumps (wire,
                         JSON_INDENT(2));
  json_decref (wire);
  GNUNET_assert (NULL != json_out);
  if (GNUNET_OK !=
      GNUNET_DISK_directory_create_for_file (ai->wire_response_filename))
  {
    GNUNET_log_strerror_file (GNUNET_ERROR_TYPE_ERROR,
                              "mkdir",
                              ai->wire_response_filename);
    global_ret = 1;
    free (json_out);
    return;
  }

  out = fopen (ai->wire_response_filename,
               "w+");
  if (NULL == out)
  {
    GNUNET_log_strerror_file (GNUNET_ERROR_TYPE_ERROR,
                              "fopen",
                              ai->wire_response_filename);
    global_ret = 1;
    free (json_out);
    return;
  }
  ret = fprintf (out,
                 "%s",
                 json_out);
  if ( (0 == fclose (out)) &&
       (-1 != ret) )
    fprintf (stdout,
             "Created wire account file `%s'\n",
             ai->wire_response_filename);
  else
    fprintf (stderr,
             "Failure creating wire account file `%s': %s\n",
             ai->wire_response_filename,
             STRERROR (errno));
  free (json_out);
}


/**
 * Main function that will be run.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  struct GNUNET_CRYPTO_EddsaPrivateKey *eddsa_priv;

  if ( (NULL == masterkeyfile) &&
       (GNUNET_OK !=
        GNUNET_CONFIGURATION_get_value_filename (cfg,
                                                 "exchange",
                                                 "MASTER_PRIV_FILE",
                                                 &masterkeyfile)) )
  {
    fprintf (stderr,
             "Master key file not given in neither configuration nor command-line\n");
    global_ret = 1;
    return;
  }
  if (GNUNET_YES !=
      GNUNET_DISK_file_test (masterkeyfile))
    GNUNET_log (GNUNET_ERROR_TYPE_INFO,
                "Exchange master private key `%s' does not exist yet, creating it!\n",
                masterkeyfile);
  eddsa_priv = GNUNET_CRYPTO_eddsa_key_create_from_file (masterkeyfile);
  if (NULL == eddsa_priv)
  {
    fprintf (stderr,
             "Failed to initialize master key from file `%s'\n",
             masterkeyfile);
    global_ret = 1;
    return;
  }
  master_priv.eddsa_priv = *eddsa_priv;
  TALER_EXCHANGEDB_find_accounts (cfg,
                                  &sign_account_data,
                                  NULL);
  GNUNET_free (eddsa_priv);
}


/**
 * The main function of the taler-exchange-wire tool.  This tool is
 * used to sign the bank account details using the master key.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc,
      char *const *argv)
{
  const struct GNUNET_GETOPT_CommandLineOption options[] = {
    GNUNET_GETOPT_option_filename ('m',
                                   "master-key",
                                   "FILENAME",
                                   "master key file (private key)",
                                   &masterkeyfile),
    GNUNET_GETOPT_OPTION_END
  };

  /* force linker to link against libtalerutil; if we do
     not do this, the linker may "optimize" libtalerutil
     away and skip #TALER_OS_init(), which we do need */
  (void) TALER_project_data_default ();
  GNUNET_assert (GNUNET_OK ==
                 GNUNET_log_setup ("taler-exchange-wire",
                                   "WARNING",
                                   NULL));
  if (GNUNET_OK !=
      GNUNET_PROGRAM_run (argc, argv,
			  "taler-exchange-wire",
			  "Setup /wire response",
			  options,
			  &run, NULL))
    return 1;
  return global_ret;
}


/* end of taler-exchange-wire.c */
